pandas/io/rdata.py

```diff
@@ -0,0 +1,588 @@
+        self.file_format = file_format.lower()
+        self.rda_name = rda_name
+        self.index = index
+        self.compression = compression
+        self.storage_options = storage_options
+
+    def verify_params(self) -> None:
+        """
+        Verify user entries of parameters.
+
+        This method will check the values and types of select parameters
+        and raise appropriate errors.
+        """
+
+        if self.file_format not in ["infer", "rda", "rdata", "rds"]:
+            raise ValueError(
+                f"{self.file_format} is not a valid value for file_format."
+            )
+
+        if (
+            self.file_format == "infer"
+            and isinstance(self.path_or_buffer, str)
+            and not self.path_or_buffer.lower().endswith((".rda", ".rdata", ".rds"))
+        ):
+            raise ValueError(
+                f"Unable to infer file format from file name: {self.path_or_buffer}"
+                "Please use known R data type (.rda, .rdata, .rds)."
+            )
+
+        if self.file_format == "infer" and isinstance(self.path_or_buffer, str):
+            self.file_format = os.path.splitext(self.path_or_buffer.lower())[1][1:]
+
+        if self.compression is not None and self.compression not in [
+            "gzip",
+            "bz2",
+            "xz",
+        ]:
+            raise ValueError(
+                f"{self.compression} is not a supported value for compression."
+            )
+
+    def disk_to_buffer(self, r_file: str) -> None:
+        """
+        Save temp file to path or buffer.
+
+        This method will convert written R data to path_or_buffer.
+        """
+
+        with open(r_file, "rb") as rdata:
+            with get_handle(
+                self.path_or_buffer,
+                "wb",
+                compression=self.compression,
+                storage_options=self.storage_options,
+                is_text=False,
+            ) as handles:
+                handles.handle.write(rdata.read())  # type: ignore[arg-type]
+
+        return None
+
+    def write_data(self) -> None:
+        """
+        Write DataFrames to R data files.
+
+        This method will run engine methods to export DataFrames
+        to R data files.
+        """
+
+        raise AbstractMethodError(self)
+
+
+class PyReadrWriter(RDataWriter):
+    """
+    Main class called in `pandas.core.frame` to write DataFrame to R
+    data types using third-party package, pyreadr.
+    """
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.verify_params()
+
+    def write_data(self) -> None:
+        from pyreadr import (
+            write_rdata,
+            write_rds,
+        )
+
+        self.frame = (
+            self.frame.reset_index()
+            if self.index
+            else self.frame.reset_index(drop=True)
+        )
+
+        with TemporaryDirectory() as tmp_dir:
+            r_temp = os.path.join(tmp_dir, "rdata.rda")
```

Comment:
Could that create problems when multiple users try to read/write R data (or the same user in multiple processes)?

Reply:
I believe `TempDirectory` runs on users' local machine and whichever user saves to final path last will see those changes. According to Python docs, [TempDirectory](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory) runs same rules as `tempfile.mkdtemp`:

> Creates a temporary directory in the most secure manner possible. There are no race conditions in the directory’s creation. The directory is readable, writable, and searchable only by the creating user ID.

Consequently, no other user will have access to that directory and files during processing. Also, paths are pretty unique like for Unix: i.e., `/tmp/tmpXKJXHZ` to indicate each call points to different temp directory.

Reply:
okay, perfect! I thought it would just return `/tmp`.